### PR TITLE
feat(components): [form-item]: add `label-position` prop

### PR DESCRIPTION
**Pull request by xing403**
Please make sure these boxes are checked before submitting your PR, thank you!

- [ ] Make sure you follow contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
- [ ] Make sure you are merging your commits to `dev` branch.
- [ ] Add some descriptions and refer to relative issues for your PR.

可对单个 form-item 进行label 定位
image
image
image

### CODE REVIEW

**Comment by xing403** [comment]
这是使用的例子

```vue
<template>
  <el-form-item label="form label-position" label-position="top">
    <el-radio-group v-model="labelPosition" size="large">
      <el-radio-button v-for="position in options" :key="position" :label="position" :value="position" />
    </el-radio-group>
  </el-form-item>
  <el-form :model="form" label-width="100px" :label-position="labelPosition" style="width: 500px">
    <el-form-item label="top" label-position="top">
      <el-input v-model="form.name" />
    </el-form-item>
    <el-form-item label="default">
      <el-col :span="11">
        <el-date-picker v-model="form.date1" type="date" placeholder="Pick a date" style="width: 100%" />
      </el-col>
      <el-col :span="2" class="text-center">
        <span class="text-gray-500">-</span>
      </el-col>
      <el-col :span="11">
        <el-time-picker v-model="form.date2" placeholder="Pick a time" style="width: 100%" />
      </el-col>
    </el-form-item>
    <el-form-item label="left" label-position="left">
      <el-radio-group v-model="form.resource">
        <el-radio value="Sponsor">Sponsor</el-radio>
        <el-radio value="Venue">Venue</el-radio>
      </el-radio-group>
    </el-form-item>
    <el-form-item label="right" label-position="right">
      <el-input v-model="form.desc" type="textarea" />
    </el-form-item>
  </el-form>
</template>

<script setup lang="ts">
import { reactive, ref } from 'vue'
let labelPosition = ref<'left' | 'right' | 'top'>('left')
const options: Array<'left' | 'right' | 'top'> = ['left', 'right', 'top']
const form = reactive({
  name: '',
  date1: '',
  date2: '',
  resource: '',
  desc: '',
})
</script>
```

**Comment by xxholly32** [comment]
https://github.com/element-plus/element-plus/discussions/16927

之前我也有提过相关讨论~

File: packages/components/form/src/form-item.vue [hunk 1]
```diff
@@ -101,7 +101,8 @@ let initialValue: any = undefined
 let isResettingField = false
 
 const labelStyle = computed<CSSProperties>(() => {
-  if (formContext?.labelPosition === 'top') {
+  const labelPosition = props?.labelPosition || formContext?.labelPosition
```
Comment (btea) [inline packages/components/form/src/form-item.vue:104]: `props?.labelPosition` optional chaining should be removed.

File: packages/components/form/src/form-item.vue [hunk 2]
```diff
@@ -135,7 +137,11 @@ const formItemClasses = computed(() => [
   formContext?.requireAsteriskPosition === 'right'
     ? 'asterisk-right'
     : 'asterisk-left',
-  { [ns.m('feedback')]: formContext?.statusIcon },
+  {
+    [ns.m('feedback')]: formContext?.statusIcon,
+    [ns.m(`label-${props?.labelPosition || formContext?.labelPosition}`)]:
+      props.labelPosition || formContext?.labelPosition,
```
Comment (btea) [inline packages/components/form/src/form-item.vue:143]: `props.labelPosition || formContext?.labelPosition` is used in multiple places and can be extracted into a computed value.

File: packages/components/form/src/form-item.ts [hunk 3]
```diff
@@ -27,6 +27,14 @@ export const formItemProps = buildProps({
     type: [String, Number],
     default: '',
   },
+  /**
+   * @description Position of label. If set to `'left'` or `'right'`, `label-width` prop is also required. The default is extend from `form label-position`.
+   */
+  labelPosition: {
```
Comment (btea) [inline packages/components/form/src/form-item.ts:33]: New properties should be added simultaneously in the documentation.

**Comment by xing403** [comment]
> #16927
> 
> 之前我也有提过相关讨论~

我在开发中也遇到了这个需求，觉得这种场景也挺常见的

File: docs/en-US/component/form.md [hunk 4]
```diff
@@ -121,6 +121,16 @@ form/accessibility
 
 :::
 
+## Custom Form Item labelPosition
```
Comment (btea) [inline docs/en-US/component/form.md:124]: Please add a version tag.
```suggestion
## Custom Form Item labelPosition ^(2.7.6)
```

File: docs/en-US/component/form.md [hunk 5]
```diff
@@ -175,6 +185,7 @@ form/accessibility
 | --------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------- | --------------------------------------------------- | ------- |
 | prop            | A key of `model`. It could be a path of the property (e.g `a.b.0` or `['a', 'b', '0']`). In the use of `validate` and `resetFields` method, the attribute is required. | ^[string] / ^[string&#91;&#93;]                     | —       |
 | label           | Label text.                                                                                                                                                            | ^[string]                                           | —       |
+| label-position  | Position of item label. If set to `'left'` or `'right'`, `label-width` prop is also required. Default extend `label-postion` of `form`                                 | ^[enum]`'left' \| 'right' \| 'top'`                 | ''      |
```
Comment (btea) [inline docs/en-US/component/form.md:188]:
```suggestion
| label-position ^(2.7.6)  | Position of item label. If set to `'left'` or `'right'`, `label-width` prop is also required. Default extend `label-postion` of `form`                                 | ^[enum]`'left' \| 'right' \| 'top'`                 | ''      |
```

File: docs/en-US/component/form.md [hunk 6]
```diff
@@ -175,6 +185,7 @@ form/accessibility
 | --------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------- | --------------------------------------------------- | ------- |
 | prop            | A key of `model`. It could be a path of the property (e.g `a.b.0` or `['a', 'b', '0']`). In the use of `validate` and `resetFields` method, the attribute is required. | ^[string] / ^[string&#91;&#93;]                     | —       |
 | label           | Label text.                                                                                                                                                            | ^[string]                                           | —       |
+| label-position ^(2.7.6)  | Position of item label. If set to `'left'` or `'right'`, `label-width` prop is also required. Default extend `label-postion` of `form`                                 | ^[enum]`'left' \| 'right' \| 'top'`                 | ''      |
```
Comment (btea) [inline docs/en-US/component/form.md:188]:
```suggestion
| label-position ^(2.7.6)  | Position of item label. If set to `'left'` or `'right'`, `label-width` prop is also required. Default extend `label-postion` of `form`.                                 | ^[enum]`'left' \| 'right' \| 'top'`                 | ''      |
```
The document should be formatted to align with the other fields.

**Review by btea — APPROVED** [review]
LGTM. Thanks!

**Review by FrontEndDog — APPROVED** [review]
Thanks for your contribution！

**Comment by kooriookami** [comment]
Could you add a complex demo, e.g. [demo](https://element-plus.run/#eyJzcmMvQXBwLnZ1ZSI6Ijx0ZW1wbGF0ZT5cbjxkaXY+XG48c3Bhbj5Gb3JtIEFsaWduPC9zcGFuPlxuICA8ZWwtcmFkaW8tZ3JvdXAgdi1tb2RlbD1cImxhYmVsUG9zaXRpb25cIiBhcmlhLWxhYmVsPVwibGFiZWwgcG9zaXRpb25cIj5cbiAgICA8ZWwtcmFkaW8tYnV0dG9uIHZhbHVlPVwibGVmdFwiPkxlZnQ8L2VsLXJhZGlvLWJ1dHRvbj5cbiAgICA8ZWwtcmFkaW8tYnV0dG9uIHZhbHVlPVwicmlnaHRcIj5SaWdodDwvZWwtcmFkaW8tYnV0dG9uPlxuICAgIDxlbC1yYWRpby1idXR0b24gdmFsdWU9XCJ0b3BcIj5Ub3A8L2VsLXJhZGlvLWJ1dHRvbj5cbiAgPC9lbC1yYWRpby1ncm91cD5cbjwvZGl2PlxuPGRpdj5cbjxzcGFuPkZpcnN0IEZvcm0gSXRlbSBBbGlnbjwvc3Bhbj5cbiAgPGVsLXJhZGlvLWdyb3VwIHYtbW9kZWw9XCJpdGVtTGFiZWxQb3NpdGlvblwiIGFyaWEtbGFiZWw9XCJsYWJlbCBwb3NpdGlvblwiPlxuICAgIDxlbC1yYWRpby1idXR0b24gdmFsdWU9XCJsZWZ0XCI+TGVmdDwvZWwtcmFkaW8tYnV0dG9uPlxuICAgIDxlbC1yYWRpby1idXR0b24gdmFsdWU9XCJyaWdodFwiPlJpZ2h0PC9lbC1yYWRpby1idXR0b24+XG4gICAgPGVsLXJhZGlvLWJ1dHRvbiB2YWx1ZT1cInRvcFwiPlRvcDwvZWwtcmFkaW8tYnV0dG9uPlxuICA8L2VsLXJhZGlvLWdyb3VwPlxuPC9kaXY+XG4gIDxkaXYgc3R5bGU9XCJtYXJnaW46IDIwcHhcIiAvPlxuICA8ZWwtZm9ybVxuICAgIDpsYWJlbC1wb3NpdGlvbj1cImxhYmVsUG9zaXRpb25cIlxuICAgIGxhYmVsLXdpZHRoPVwiMTUwcHhcIlxuICAgIDptb2RlbD1cImZvcm1MYWJlbEFsaWduXCJcbiAgICBzdHlsZT1cIm1heC13aWR0aDogNjAwcHhcIlxuICA+XG4gICAgPGVsLWZvcm0taXRlbSBsYWJlbD1cIk5hbWVcIiA6bGFiZWwtcG9zaXRpb249XCJpdGVtTGFiZWxQb3NpdGlvblwiPlxuICAgICAgPGVsLWlucHV0IHYtbW9kZWw9XCJmb3JtTGFiZWxBbGlnbi5uYW1lXCIgLz5cbiAgICA8L2VsLWZvcm0taXRlbT5cbiAgICA8ZWwtZm9ybS1pdGVtIGxhYmVsPVwiQWN0aXZpdHkgem9uZVwiPlxuICAgICAgPGVsLWlucHV0IHYtbW9kZWw9XCJmb3JtTGFiZWxBbGlnbi5yZWdpb25cIiAvPlxuICAgIDwvZWwtZm9ybS1pdGVtPlxuICAgIDxlbC1mb3JtLWl0ZW0gbGFiZWw9XCJBY3Rpdml0eSBmb3JtXCI+XG4gICAgICA8ZWwtaW5wdXQgdi1tb2RlbD1cImZvcm1MYWJlbEFsaWduLnR5cGVcIiAvPlxuICAgIDwvZWwtZm9ybS1pdGVtPlxuICA8L2VsLWZvcm0+XG48L3RlbXBsYXRlPlxuXG48c2NyaXB0IGxhbmc9XCJ0c1wiIHNldHVwPlxuaW1wb3J0IHsgcmVhY3RpdmUsIHJlZiB9IGZyb20gJ3Z1ZSdcbmltcG9ydCB0eXBlIHsgRm9ybVByb3BzIH0gZnJvbSAnZWxlbWVudC1wbHVzJ1xuXG5jb25zdCBsYWJlbFBvc2l0aW9uID0gcmVmPEZvcm1Qcm9wc1snbGFiZWxQb3NpdGlvbiddPigncmlnaHQnKVxuY29uc3QgaXRlbUxhYmVsUG9zaXRpb24gPSByZWY8Rm9ybVByb3BzWydsYWJlbFBvc2l0aW9uJ10+KCdyaWdodCcpXG5cbmNvbnN0IGZvcm1MYWJlbEFsaWduID0gcmVhY3RpdmUoe1xuICBuYW1lOiAnJyxcbiAgcmVnaW9uOiAnJyxcbiAgdHlwZTogJycsXG59KVxuPC9zY3JpcHQ+XG4iLCJpbXBvcnQtbWFwLmpzb24iOiJ7XG4gIFwiaW1wb3J0c1wiOiB7XG4gICAgXCJlbGVtZW50LXBsdXNcIjogXCJodHRwczovL3ByZXZpZXctMTcxMTEtZWxlbWVudC1wbHVzLnN1cmdlLnNoL2J1bmRsZS9pbmRleC5mdWxsLm1pbi5tanNcIixcbiAgICBcImVsZW1lbnQtcGx1cy9cIjogXCJ1bnN1cHBvcnRlZFwiXG4gIH1cbn0iLCJ0c2NvbmZpZy5qc29uIjoie1xuICBcImNvbXBpbGVyT3B0aW9uc1wiOiB7XG4gICAgXCJ0YXJnZXRcIjogXCJFU05leHRcIixcbiAgICBcImpzeFwiOiBcInByZXNlcnZlXCIsXG4gICAgXCJtb2R1bGVcIjogXCJFU05leHRcIixcbiAgICBcIm1vZHVsZVJlc29sdXRpb25cIjogXCJCdW5kbGVyXCIsXG4gICAgXCJ0eXBlc1wiOiBbXCJlbGVtZW50LXBsdXMvZ2xvYmFsLmQudHNcIl0sXG4gICAgXCJhbGxvd0ltcG9ydGluZ1RzRXh0ZW5zaW9uc1wiOiB0cnVlLFxuICAgIFwiYWxsb3dKc1wiOiB0cnVlLFxuICAgIFwiY2hlY2tKc1wiOiB0cnVlXG4gIH0sXG4gIFwidnVlQ29tcGlsZXJPcHRpb25zXCI6IHtcbiAgICBcInRhcmdldFwiOiAzLjNcbiAgfVxufVxuIiwiX28iOnsic3R5bGVTb3VyY2UiOiJodHRwczovL3ByZXZpZXctMTcxMTEtZWxlbWVudC1wbHVzLnN1cmdlLnNoL2J1bmRsZS9pbmRleC5jc3MifX0=)

May be it can be more prettier.

File: docs/en-US/component/form.md [hunk 7]
```diff
@@ -121,6 +121,16 @@ form/accessibility
 
 :::
 
+## Custom Form Item labelPosition ^(2.7.6)
+
+The label of the component can be positioned using the `label-position` property in el-form-item. The default extend `el-form`
```
Comment (kooriookami) [inline docs/en-US/component/form.md:126]:
```suggestion
You can set the position of `el-form-item` separately using `label-position`. If the value is not set, the `label-position` of `el-form` is used.
```

**Review by kooriookami — CHANGES_REQUESTED** [review]

**Comment by kooriookami** [comment]
image

May be we can merge the two demos in one demo.

**Comment by xing403** [comment]
> image
> 
> May be we can merge the two demos in one demo.

好的好的

File: docs/en-US/component/form.md [hunk 8]
```diff
@@ -168,33 +158,33 @@ form/form-item-labelPosition
...
+| ----------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------- | --------------------------------------------------- | ------- |
+| prop                    | A key of `model`. It could be a path of the property (e.g `a.b.0` or `['a', 'b', '0']`). In the use of `validate` and `resetFields` method, the attribute is required. | ^[string] / ^[string&#91;&#93;]                     | —       |
+| label                   | Label text.                                                                                                                                                            | ^[string]                                           | —       |
+| label-position ^(2.7.6) | Position of item label. If set to `'left'` or `'right'`, `label-width` prop is also required. Default extend `label-postion` of `form`.                                | ^[enum]`'left' \| 'right' \| 'top'`                 | ''      |
```
Comment (kooriookami) [inline docs/en-US/component/form.md:178]:
```suggestion
| label-position ^(2.7.7) | Position of item label. If set to `'left'` or `'right'`, `label-width` prop is also required. Default extend `label-postion` of `form`.                                | ^[enum]`'left' \| 'right' \| 'top'`                 | ''      |
```

File: docs/en-US/component/form.md [hunk 9]
```diff
@@ -45,7 +45,7 @@ form/inline-form
 
 ## Alignment
 
-Depending on your design, there are several different ways to align your label element.
+Depending on your design, there are several different ways to align your label element. You can set the position of `el-form-item` separately using `label-position`. If the value is not set, the `label-position` of `el-form` is used.
```
Comment (kooriookami) [inline docs/en-US/component/form.md:48]:
```suggestion
Depending on your design, there are several different ways to align your label element. You can set the position of `el-form-item` separately using `label-position` ^(2.7.7). If the value is not set, the `label-position` of `el-form` is used.
```

**Review by kooriookami — CHANGES_REQUESTED** [review]

**Review by kooriookami — APPROVED** [review]
LGTM

**Pull request merged**